src/MySqlConnector/MySqlClient/ConnectionPool.cs

```diff
@@ -91,15 +85,47 @@ private bool SessionIsHealthy(MySqlSession session)
 
 		public void Return(MySqlSession session)
 		{
+			// NOTE: deliberately calling 'async void' method so that first part runs synchronously, but remainder gets queued
```

Comment:
No, this is definitely an anti-pattern:

* [Async/Await - Best Practices in Asynchronous Programming](https://msdn.microsoft.com/en-us/magazine/jj991977.aspx) - "avoid async void"
* [Avoid async void methods](http://haacked.com/archive/2014/11/11/async-void-methods/)

With `async Task` one has to explicitly disable a compiler warning:

```
#pragma warning disable 4014
    ReturnAsync(session);
#pragma warning restore 4014
```

The final option is `Task.Run(async () => await ReturnAsync(session));`. I avoided this because I wanted the first part to run synchronously without being queued to the threadpool. I also wanted to avoid extra allocations from the lambda and extra `Task`. (The comment didn't make this clear, but should have.)